src/messages/direct.rs

```diff
@@ -0,0 +1,234 @@
+                part_index,
+                part_count,
+                ref proof,
+                leading_zero_bytes,
+            } => write!(
+                formatter,
+                "ResourceProofResponse {{ part {}/{}, proof_len: {:?}, leading_zero_bytes: \
+                 {:?} }}",
+                part_index + 1,
+                part_count,
+                proof.len(),
+                leading_zero_bytes
+            ),
+            ResourceProofResponseReceipt => write!(formatter, "ResourceProofResponseReceipt"),
+            ProxyRateLimitExceeded { ref ack } => {
+                write!(formatter, "ProxyRateLimitExceeded({:?})", ack)
+            }
+            ParsecRequest(ref v, _) => write!(formatter, "ParsecRequest({}, _)", v),
+            ParsecResponse(ref v, _) => write!(formatter, "ParsecResponse({}, _)", v),
+            ParsecPoke(ref v) => write!(formatter, "ParsecPoke({})", v),
+        }
+    }
+}
+
+#[cfg_attr(feature = "mock_serialise", derive(Clone))]
+#[derive(Serialize, Deserialize)]
+pub struct SignedDirectMessage {
+    content: DirectMessage,
+    src_id: PublicId,
+    signature: Signature,
+}
+
+impl SignedDirectMessage {
+    /// Create new `DirectMessage` with `content` and signed by `src_full_id`.
+    pub fn new(content: DirectMessage, src_full_id: &FullId) -> Result<Self, RoutingError> {
+        let signature = self::implementation::sign(src_full_id, &content)?;
+
+        Ok(Self {
+            content,
+            src_id: *src_full_id.public_id(),
+            signature,
+        })
+    }
+
+    /// Verify the message signature.
+    pub fn verify(&self) -> Result<(), RoutingError> {
+        self::implementation::verify(&self.src_id, &self.signature, &self.content)
+    }
+
+    /// The priority Crust should send this message with.
+    pub fn priority(&self) -> u8 {
+        match self.content {
+            DirectMessage::ResourceProofResponse { .. } => 9,
+            _ => 0,
+        }
+    }
+
+    /// Verify the message signature and return its content and the sender id.
+    /// Consume the message in the process.
+    pub fn open(self) -> Result<(DirectMessage, PublicId), RoutingError> {
+        self.verify()?;
+        Ok((self.content, self.src_id))
+    }
+
+    /// Content of the message.
+    #[cfg(any(test, feature = "mock_serialise"))]
+    pub fn content(&self) -> &DirectMessage {
+        &self.content
+    }
+}
+
+impl Debug for SignedDirectMessage {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(
+            formatter,
+            "SignedDirectMessage {{ content: {:?}, src_id: {:?}, signature: {:?} }}",
+            self.content, self.src_id, self.signature
+        )
+    }
+}
+
+#[cfg(not(feature = "mock_serialise"))]
+mod implementation {
+    use super::*;
+    use maidsafe_utilities::serialisation;
+
+    pub fn sign(src_full_id: &FullId, content: &DirectMessage) -> Result<Signature, RoutingError> {
```

Comment:
I think I would rename this to `sign_public_id_and_content` because the signature as is could be taken as: "sign this content with the private key from this FullId". Being more explicit would make the call site unambiguous.

Reply:
Not sure I get you completely, but signing the public id is important to know the sender is authentic. Singing the content is something I decided to add, because it was easy to do and I think it provides additional layer of security we didn't have before this PR (previously `DirectMessage` weren't signed). For example, if only the public id were signed, not the content, then an advresary can in theory capture a legit message, take the public_id and the signature, put their own content and pass it on and the recipient would have no way of knowing the message was tampered with.

Reply:
My point is that just signing the content would probably be enough (so we could name the function `sign_content` or something) - the public ID is just a piece of data that enables validation of the signature.

Basically, if the signature of the content is valid when we use the given public ID, it already means that someone with the corresponding private key signed it - no need to sign the ID, too.

Reply:
I see. Yeah, that makes sense. I'll update the code.